File: 2020/900.py
"""
see https://leetcode.com/problems/rle-iterator/
"""
class RLEIterator(object):

    def __init__(self, A):
        """
        :type A: List[int]
        """
        self.rpt = [a for a in A[::2]]
        self.num = [a for a in A[1::2]]
        self.pos = 0
        # print(self.rpt)
        # print(self.num)

    def next(self, n):
        """
        :type n: int
        :rtype: int
        """
        i = self.pos
        while n:
            if i >= len(self.rpt):
                self.pos = i
                return -1
            if n-self.rpt[i] > 0:
                n, self.rpt[i] = n-self.rpt[i], 0
                i += 1
            else:
                n, self.rpt[i] = 0, self.rpt[i]-n

        self.pos = i
        # print(self.pos)

        return self.num[i]

# Your RLEIterator object will be instantiated and called as such:
# obj = RLEIterator(A)
# param_1 = obj.next(n)

File: current_session/c++/900.cpp
#include <iostream>

using namespace std;

// simply use a counter and return the current value

class RLEIterator {
public:
  int cnt = 0;
  int eIndex = 1;
  vector<int> encodingArr;
  RLEIterator(vector<int>& encoding) {
    encodingArr = encoding;
  }
  
  int next(int n) {
    // get the answer while maintain the array
    int ans = -1;

    // keep decreasing n
    // move to next index (also consider cnt == encodingArr[eIndex-1])
    while (eIndex < encodingArr.size() && cnt+n > encodingArr[eIndex-1] ) {
      n -= (encodingArr[eIndex-1] - cnt);
      cnt = 0;
      eIndex += 2;
    }
    
    // get the answer
    if (eIndex < encodingArr.size() && cnt+n <= encodingArr[eIndex-1]) {
      cnt += n;
      ans = encodingArr[eIndex];
      n = 0;
    }
   
    return ans;
  }
};

/**
 * Your RLEIterator object will be instantiated and called as such:
 * RLEIterator* obj = new RLEIterator(encoding);
 * int param_1 = obj->next(n);
 */